src/roompanel/RoomManager.cpp

```diff
@@ -0,0 +1,252 @@
+                                 << "containing invalid Json: expecting array, got"
+                                 << msg.getJson().toQString();
+        }
+        return;
+    }
+    m_room->resetMobs();
+    for (QJsonValueRef value : doc.array()) {
+        if (value.isObject()) {
+            addMob(value.toObject());
+        } else {
+            if (false) {
+                qWarning().noquote()
+                    << "RoomManager received GMCP" << msg.getName().toQString()
+                    << "containing invalid Json: expecting array of objects, got [/*...*/"
+                    << value.type() << "/*...*/]";
+            }
+        }
+    }
+    updateWidget();
+}
+
+void RoomManager::parseGmcpUpdate(const GmcpMessage &msg)
+{
+    showGmcp(msg);
+    if (isInvalidJson(msg)) {
+        return;
+    }
+    const QJsonDocument &doc = msg.getDocument();
+    if (!doc.isObject()) {
+        if (false) {
+            qWarning().noquote() << "RoomManager received GMCP" << msg.getName().toQString()
+                                 << "containing invalid Json: expecting object, got"
+                                 << msg.getJson().toQString();
+        }
+        return;
+    }
+    updateMob(doc.object());
+}
+
+void RoomManager::addMob(QJsonObject obj)
+{
+    RoomMobUpdate data;
+    if (toMob(obj, data)) {
+        m_room->addMob(std::move(data));
+        updateWidget();
+    }
+}
+
+void RoomManager::updateMob(QJsonObject obj)
+{
+    RoomMobUpdate data;
+    if (toMob(obj, data)) {
+        if (m_room->updateMob(std::move(data))) {
+            updateWidget();
+        }
+    }
+}
+
+bool RoomManager::isInvalidJson(const GmcpMessage &msg)
+{
+    if (msg.hasError()) {
+        if (false) {
+            qWarning().noquote() << "RoomManager received GMCP containing invalid Json:"
+                                 << msg.getError().errorString();
+        }
+        return true;
+    }
+    return false;
+}
+
+void RoomManager::showGmcp(const GmcpMessage &msg)
+{
+    if (false) {
+        qDebug().noquote() << "RoomManager received GMCP:" << msg.toRawBytes();
+    }
+}
+
+const QHash<QString, MobFieldEnum> RoomManager::mobFields{
+    {"name", MobFieldEnum::NAME},
+    {"desc", MobFieldEnum::DESC},
+    {"fighting", MobFieldEnum::FIGHTING},
+    {"flags", MobFieldEnum::FLAGS},
+    {"labels", MobFieldEnum::LABELS},
+    {"riding", MobFieldEnum::MOUNT},
+    {"driving", MobFieldEnum::MOUNT},
+    {"position", MobFieldEnum::POSITION},
+    {"weapon", MobFieldEnum::WEAPON},
+};
+
+bool RoomManager::toMob(QJsonObject obj, RoomMobUpdate &data)
+{
+    QJsonObject::iterator iter = obj.find("id");
+    QJsonObject::iterator end = obj.end();
+    if (iter == end || !toMobId(*iter, data.id())) {
+        return false;
+    }
+    auto matchEnd = mobFields.end();
+    for (iter = obj.begin(); iter != end; ++iter) {
+        auto match = mobFields.find(iter.key());
+        if (match != matchEnd) {
```

Comment:
What's the impact of this being matchEnd? Should we handle this?

Reply:
All JSON object keys are optional in GMCP messages Room.Chars.*  - only 'id' is mandatory.

If `match == matchEnd` it means one of the JSON object keys we know about (name, desc, fighting, flags, labels, riding, driving, position, weapon) is not present in the GMCP message, so we don't save the corresponding value into `RoomMobUpdate`

Reply:
I have added
```C++
if (m_debug) {
  qWarning() << // ...
}
```